### PR TITLE
feat: add extensions.kubesphere.io CRDs

### PR DESCRIPTION
**Pull request by wansir**
### What type of PR is this?

/kind feature

### What this PR does / why we need it:

The first PR of KubeSphere pluggable architecture. Add CRDs in API group extensions.kubesphere.io.  

### Does this PR introduced a user-facing change?

```release-note
None
```

### Additional documentation, usage docs, etc.:

Design docs in Chinese https://github.com/kubesphere/community/blob/master/sig-architecture/concepts-and-designs/pluggable-architecture-zh.md

### CODE REVIEW

File: staging/src/kubesphere.io/api/extensions/v1alpha1/plugin_types.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+	Status PluginStatus `json:"status,omitempty"`
+}
+
+type UpdateStrategy struct {
+	*RegistryPoll `json:"registryPoll,omitempty"`
+}
+
+type RegistryPoll struct {
+	Interval *metav1.Duration `json:"interval,omitempty"`
```
Comment (xyz-li) [inline staging/src/kubesphere.io/api/extensions/v1alpha1/plugin_types.go:94]: RawInterval  string           `json:"interval,omitempty"`

The underlying type of metav1.Duration is int64. We should add a string interval to facilitate the user config.

File: staging/src/kubesphere.io/api/extensions/v1alpha1/plugin_types.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:categories="plugin",scope="Cluster"
+
+// Plugin is synchronized from the Repository.
+// A plugin can contain multiple versions.
+type Plugin struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
```
Comment (xyz-li) [inline staging/src/kubesphere.io/api/extensions/v1alpha1/plugin_types.go:85]: Is there no `Spec` field?

File: staging/src/kubesphere.io/api/extensions/v1alpha1/plugin_types.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	*RegistryPoll `json:"registryPoll,omitempty"`
+}
+
+type RegistryPoll struct {
+	Interval *metav1.Duration `json:"interval,omitempty"`
+}
+
+type RepositorySpec struct {
+	Image          string         `json:"image"`
+	DisplayName    string         `json:"displayName,omitempty"`
```
Comment (xyz-li) [inline staging/src/kubesphere.io/api/extensions/v1alpha1/plugin_types.go:99]: Should we add a `Description` field to show more specific information about this repository?

File: staging/src/kubesphere.io/api/extensions/v1alpha1/plugin_types.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+// Category can help us group the plugins.
+type Category struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              CategorySpec `json:"spec,omitempty"`
+}
+
+type PluginStatus struct {
+	Phase    string   `json:"phase"`
+	Versions []string `json:"versions,omitempty"`
```
Comment (xyz-li) [inline staging/src/kubesphere.io/api/extensions/v1alpha1/plugin_types.go:67]: If there are multiple versions for one plugin. Is there a default version for this plugin or not? Should we add a description for this plugin?

File: staging/src/kubesphere.io/api/extensions/v1alpha1/plugin_types.go (hunk 1 above)
Comment (wansir) [reply staging/src/kubesphere.io/api/extensions/v1alpha1/plugin_types.go:94]: `metav1.Duration` is able to unmarshaled from string,  `RawInterval` is unnecessary.

File: staging/src/kubesphere.io/api/extensions/v1alpha1/plugin_types.go (hunk 3 above)
Comment (wansir) [reply staging/src/kubesphere.io/api/extensions/v1alpha1/plugin_types.go:99]: Good suggestion.

File: staging/src/kubesphere.io/api/extensions/v1alpha1/subscription_types.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+type PluginRef struct {
+	Name    string `json:"name"`
+	Version string `json:"version"`
+}
+
+type SubscriptionSpec struct {
+	Plugin          PluginRef            `json:"plugin"`
+	Enabled         bool                 `json:"enabled"`
+	TargetNamespace string               `json:"targetNamespace"`
```
Comment (xyz-li) [inline staging/src/kubesphere.io/api/extensions/v1alpha1/subscription_types.go:32]: Can the user specify the name of the release?

Comment (xyz-li) [reply staging/src/kubesphere.io/api/extensions/v1alpha1/subscription_types.go:32]: How do we generate a release name for a release?

File: staging/src/kubesphere.io/api/extensions/v1alpha1/plugin_types.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	Email string `json:"email,omitempty"`
+	Name  string `json:"name,omitempty"`
+	URL   string `json:"url,omitempty"`
+}
+
+type PluginInfo struct {
+	Description string       `json:"description,omitempty"`
+	Icon        string       `json:"icon,omitempty"`
+	Maintainers []Maintainer `json:"maintainers,omitempty"`
+	Version     string       `json:"version,omitempty"`
```
Comment (xyz-li) [inline staging/src/kubesphere.io/api/extensions/v1alpha1/plugin_types.go:32]: Should we add a digest field to check whether the content of the plugin changed? The user may have changed the content of the plugin, but forgot to change the version.

File: staging/src/kubesphere.io/api/extensions/v1alpha1/apiservice_types.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+type APIServiceSpec struct {
+	Group              string   `json:"group,omitempty"`
+	Version            string   `json:"version,omitempty"`
+	InsecureSkipVerify bool     `json:"insecureSkipVerify,omitempty"`
+	NonResourceURLs    []string `json:"nonResourceURLs,omitempty"`
+	Endpoint           string   `json:"endpoint,omitempty"`
```
Comment (xyz-li) [inline staging/src/kubesphere.io/api/extensions/v1alpha1/apiservice_types.go:26]: Why not define service like this [ServiceReference ](https://github.com/kubernetes/apiextensions-apiserver/blob/24b17b094a2fb8032e6d168ac7c400184d414527/pkg/apis/apiextensions/types.go#L151-L169)
```
// ServiceReference holds a reference to Service.legacy.k8s.io
type ServiceReference struct {
	// `namespace` is the namespace of the service.
	// Required
	Namespace string
	// `name` is the name of the service.
	// Required
	Name string


	// `path` is an optional URL path which will be sent in any request to
	// this service.
	// +optional
	Path *string


	// If specified, the port on the service that hosting webhook.
	// `port` should be a valid port number (1-65535, inclusive).
	// +optional
	Port int32
}
```

File: staging/src/kubesphere.io/api/extensions/v1alpha1/jsbundle_types.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+	// +optional
+	Raw []byte `json:"raw,omitempty"`
+	// +optional
+	RawFrom RawFrom `json:"rawFrom,omitempty"`
+	// +optional
+	FilenameOverride string `json:"filenameOverride,omitempty"`
+}
+
+type RawFrom struct {
+	URL string `json:"url,omitempty"`
```
Comment (xyz-li) [inline staging/src/kubesphere.io/api/extensions/v1alpha1/jsbundle_types.go:35]: If there is a URL the ks-apiserver need to access,  should there is a CABundule to secure the request?

File: staging/src/kubesphere.io/api/extensions/v1alpha1/plugin_types.go (hunk 2 above)
Comment (wansir) [reply staging/src/kubesphere.io/api/extensions/v1alpha1/plugin_types.go]: Updated.
image

**Pull request merged**